firebase-common/src/main/java/com/google/firebase/Timestamp.kt

```diff
@@ -0,0 +1,141 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase
+
+import android.os.Parcel
+import android.os.Parcelable
+import java.util.Date
+
+/**
+ * A Timestamp represents a point in time independent of any time zone or calendar.
+ *
+ * Represented as seconds and fractions of seconds at nanosecond resolution in UTC Epoch time. It is
+ * encoded using the Proleptic Gregorian Calendar which extends the Gregorian calendar backwards to
+ * year one. Furthermore,It is encoded assuming all minutes are 60 seconds long, specifically leap
+ * seconds are "smeared" so that no leap second table is needed for interpretation. Range is from
+ * 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z. By restricting to that range, we ensure
+ * that we can convert to and from RFC 3339 date strings.
+ *
+ * @see <a href="https://git.page.link/timestamp-proto">Timestamp</a>The ref timestamp definition
+ */
+class Timestamp : Comparable<Timestamp>, Parcelable {
+  val seconds: Long
+  val nanoseconds: Int
+
+  /**
+   * Creates a new [Timestamp].
+   *
+   * @param seconds represents seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z. Must be
+   * from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59Z inclusive.
+   * @param nanoseconds represents non-negative fractions of a second at nanosecond resolution.
+   * Negative second values with fractions must still have non-negative nanoseconds values that
+   * count forward in time. Must be from 0 to 999,999,999 inclusive.
+   */
+  constructor(seconds: Long, nanoseconds: Int) {
+    validateRange(seconds, nanoseconds)
+
+    this.seconds = seconds
+    this.nanoseconds = nanoseconds
+  }
+
+  constructor(source: Parcel) : this(source.readLong(), source.readInt())
+
+  constructor(date: Date) {
+    val (seconds, nanoseconds) = date.toPreciseTime()
+
+    validateRange(seconds, nanoseconds)
+
+    this.seconds = seconds
+    this.nanoseconds = nanoseconds
+  }
+
+  /**
+   * Returns a new [Date] corresponding to this timestamp.
+   *
+   * This may lose precision.
+   */
+  fun toDate(): Date = Date(seconds * 1_000 + (nanoseconds / 1_000_000))
+
+  override fun compareTo(other: Timestamp): Int =
+    compareValuesBy(this, other, Timestamp::seconds, Timestamp::nanoseconds)
+
+  override fun equals(other: Any?): Boolean =
+    when {
+      (other === this) -> true
+      (other !is Timestamp) -> false
+      else -> compareTo(other) == 0
+    }
+
+  override fun hashCode(): Int {
+    val prime = 37
+    val initialHash = prime * seconds.toInt()
+    val withHighOrderBits = prime * initialHash + seconds.shr(32).toInt()
+
+    return prime * withHighOrderBits + nanoseconds
+  }
+
+  override fun toString(): String = "Timestamp(seconds=$seconds, nanoseconds=$nanoseconds)"
+
+  override fun describeContents(): Int {
+    return 0
+  }
```

Comment:
Optional: Consider re-writing the implementation of `describeContents()` as just `= 0`

Reply:
sgtm!